File: src/main/java/com/hileco/cortex/context/layer/LayeredStack.java
package com.hileco.cortex.context.layer;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Objects;

public class LayeredStack<V> implements StackApi<V, LayeredStack<V>> {
    private LayeredStack<V> parent;
    private Map<Integer, V> layer;
    private int size;

    public LayeredStack() {
        this.parent = null;
        this.layer = new HashMap<>();
        this.size = 0;
    }

    private LayeredStack(LayeredStack<V> parent, Map<Integer, V> layer, int size) {
        this.parent = parent;
        this.layer = layer;
        this.size = size;
    }

    @Override
    public synchronized LayeredStack<V> copy() {
        if (this.layer.size() > 0 || this.size != this.parent.size) {
            this.parent = new LayeredStack<>(this.parent, this.layer, this.size);
            this.layer = new HashMap<>();
            this.size = this.parent.size;
        }
        return new LayeredStack<>(this.parent, new HashMap<>(), this.parent.size);
    }

    @Override
    public synchronized void push(V value) {
        this.size++;
        this.layer.put(this.size, value);
    }

    @Override
    public synchronized V pop() {
        V removed = this.layer.remove(this.size);
        this.size--;
        return removed;
    }

    public synchronized V peek() {
        return this.layer.get(this.size);
    }

    private UnsupportedOperationException arbitraryModification() {
        return new UnsupportedOperationException("Modifying elements at arbitrary positions is not supported.");
    }

    private synchronized void checkBounds(int index) {
        if (this.size < index) {
            throw new IndexOutOfBoundsException(String.format("Size %s < Index %s", this.size, index));
        }
    }

    @Override
    public synchronized V get(int index) {
        this.checkBounds(index);
        if (this.layer.containsKey(index)) {
            return this.layer.get(index);
        } else {
            if (this.parent == null) {
                throw new IndexOutOfBoundsException(String.format("Size %s < Index %s", this.size, index));
            }
            return this.parent.get(index);
        }
    }

    @Override
    public synchronized void swap(int topOffsetLeft, int topOffsetRight) {
        int indexA = this.size - topOffsetLeft;
        int indexB = this.size - topOffsetRight;
        this.checkBounds(indexA);
        this.checkBounds(indexB);
        V valueA = this.get(indexA);
        V valueB = this.get(indexB);
        this.layer.put(indexA, valueB);
        this.layer.put(indexB, valueA);
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public synchronized boolean isEmpty() {
        return this.layer.size() == 0 && (this.parent == null || this.parent.size == 0);
    }

    @Override
    public synchronized void duplicate(int topOffset) {
        int index = this.size - topOffset;
        this.checkBounds(index);
        V value = this.get(index);
        this.push(value);
    }

    @Override
    public synchronized boolean contains(Object o) {
        if (o instanceof Integer) {
            return this.layer.containsKey(o) || (this.parent != null && this.parent.contains(o));
        } else {
            throw new IllegalArgumentException();
        }
    }

    @Override
    public IndexedIterator iterator() {
        return new IndexedIterator(1);
    }

    @SuppressWarnings("unchecked")
    @Override
    public synchronized V[] toArray() {
        Object[] objects = new Object[this.size];
        for (int index = 0; index < this.size; index++) {
            V value = this.get(index + 1);
            objects[index] = value;
        }
        return (V[]) objects;
    }

    @SuppressWarnings("unchecked")
    @Override
    public synchronized <T> T[] toArray(T[] a) {
        if (a.length < this.size) {
            return (T[]) this.toArray();
        } else {
            for (int index = 0; index < a.length; index++) {
                V value = this.get(index + 1);
                a[index] = (T) value;
            }
            return a;
        }
    }

    @Override
    public synchronized boolean add(V v) {
        this.push(v);
        return true;
    }

    @Override
    public boolean remove(Object o) {
        throw this.arbitraryModification();
    }

    @Override
    public synchronized boolean containsAll(Collection<?> c) {
        return c.stream().allMatch(this::contains);
    }

    @Override
    public synchronized boolean addAll(Collection<? extends V> c) {
        c.forEach(this::push);
        return true;
    }

    @Override
    public boolean addAll(int offset, Collection<? extends V> c) {
        throw this.arbitraryModification();
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        throw this.arbitraryModification();
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        throw this.arbitraryModification();
    }

    @Override
    public synchronized void clear() {
        this.layer.clear();
        this.size = 0;
    }

    @Override
    public V set(int index, V element) {
        throw this.arbitraryModification();
    }

    @Override
    public void add(int index, V element) {
        throw this.arbitraryModification();
    }

    @Override
    public V remove(int index) {
        throw this.arbitraryModification();
    }

    @Override
    public synchronized int indexOf(Object o) {
        IndexedIterator iterator = new IndexedIterator(1);
        while (iterator.hasNext()) {
            if (Objects.equals(iterator.next(), o)) {
                return iterator.index - 1;
            }
        }
        return -1;
    }

    @Override
    public synchronized int lastIndexOf(Object o) {
        IndexedIterator iterator = new IndexedIterator(1);
        int match = -1;
        while (iterator.hasNext()) {
            if (Objects.equals(iterator.next(), o)) {
                match = iterator.index - 1;
            }
        }
        return match;
    }

    @Override
    public ListIterator<V> listIterator() {
        return new IndexedIterator(1);
    }

    @Override
    public ListIterator<V> listIterator(int index) {
        if (index < 0) {
            throw new IndexOutOfBoundsException();
        }
        if (index > this.size) {
            throw new IndexOutOfBoundsException();
        }
        return new IndexedIterator(index);
    }

    @Override
    public List<V> subList(int fromIndex, int toIndex) {
        throw new UnsupportedOperationException("Creating in-memory sub-lists is currently not supported.");
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("LayeredStack{");
        IndexedIterator iterator = this.iterator();
        while (iterator.hasNext()) {
            stringBuilder.append(iterator.next());
            if (iterator.hasNext()) {
                stringBuilder.append(", ");
            }
        }
        stringBuilder.append("}");
        return stringBuilder.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;
        LayeredStack<?> that = (LayeredStack<?>) o;
        return this.size == that.size &&
                Arrays.equals(this.toArray(), that.toArray());
    }

    @Override
    public int hashCode() {
        return Objects.hash((Object[]) this.toArray());
    }

    private class IndexedIterator implements ListIterator<V> {
        private int index;

        IndexedIterator(int index) {
            this.index = index;
        }

        @Override
        public synchronized boolean hasNext() {
            return this.index <= LayeredStack.this.size;
        }

        @Override
        public synchronized V next() {
            if (this.index > LayeredStack.this.size) {
                throw new NoSuchElementException();
            }
            V value = LayeredStack.this.get(this.index);
            this.index++;
            return value;
        }

        @Override
        public boolean hasPrevious() {
            return this.index > 0;
        }

        @Override
        public V previous() {
            if (this.index == 0) {
                throw new NoSuchElementException();
            }
            this.index--;
            return LayeredStack.this.get(this.index);
        }

        @Override
        public int nextIndex() {
            return Math.min(this.index, LayeredStack.this.size - 1);
        }

        @Override
        public int previousIndex() {
            return this.index - 1;
        }

        @Override
        public void remove() {
            throw LayeredStack.this.arbitraryModification();
        }

        @Override
        public void set(V v) {
            throw LayeredStack.this.arbitraryModification();
        }

        @Override
        public void add(V v) {
            throw LayeredStack.this.arbitraryModification();
        }
    }
}

File: src/main/java/com/hileco/cortex/vm/layer/LayeredStack.kt
package com.hileco.cortex.vm.layer

import java.math.BigInteger

class LayeredStack<V> : TreeLayered<LayeredStack<V>> {
    private var size: Int
    val layer: HashMap<Int, V>

    private constructor(parent: LayeredStack<V>?,
                        size: Int,
                        layer: HashMap<Int, V>) : super(parent) {
        this.size = size
        this.layer = layer
        val mergeableParent = this.parent
        if (mergeableParent != null && mergeableParent.layer.size < MINIMUM_LAYER_SIZE) {
            mergeParent(mergeableParent)
            this.parent = mergeableParent.parent()
        }
    }

    constructor() {
        size = 0
        layer = HashMap()
    }

    fun push(value: V) {
        layer[this.size] = value
        this.size++
    }

    fun pop(): V {
        val value = peek()
        layer.remove(this.size - 1)
        this.size--
        return value
    }

    fun peek(offset: Int = 0): V {
        return if (this.size <= offset) throw IndexOutOfBoundsException("size 0 <= index 0") else this[this.size - offset - 1]
    }

    operator fun get(index: Int): V {
        var chosen: LayeredStack<V>? = this
        while (chosen != null && chosen.size > index) {
            val value = chosen.layer[index]
            if (value != null) {
                return value
            }
            chosen = chosen.parent()
        }
        throw IndexOutOfBoundsException("size ${this.size} <= index $index")
    }

    operator fun set(index: Int, value: V) {
        if (this.size < index) {
            throw IndexOutOfBoundsException("size ${this.size} <= index $index")
        }
        layer[index] = value
    }

    fun swap(topOffsetLeft: Int, topOffsetRight: Int) {
        val leftIndex = (this.size - 1) - topOffsetLeft
        val rightIndex = (this.size - 1) - topOffsetRight
        val left = this[leftIndex]
        val right = this[rightIndex]
        layer[leftIndex] = right
        layer[rightIndex] = left
    }

    fun size(): Int {
        return this.size
    }

    fun isEmpty(): Boolean {
        return this.size == 0
    }

    fun duplicate(topOffset: Int) {
        val value = peek(topOffset)
        push(value)
    }

    fun clear() {
        layer.clear()
        this.size = 0
    }

    @Synchronized
    override fun extractParentLayer(parent: LayeredStack<V>?): LayeredStack<V> {
        val extracted = LayeredStack(parent, size, HashMap(layer))
        layer.clear()
        return extracted
    }

    @Synchronized
    override fun isLayerEmpty(): Boolean {
        return layer.isEmpty()
    }

    @Synchronized
    override fun createSibling(parent: LayeredStack<V>?): LayeredStack<V> {
        return LayeredStack(parent, parent?.size ?: 0, HashMap())
    }

    @Synchronized
    override fun mergeParent(parent: LayeredStack<V>) {
        synchronized(parent) {
            parent.layer.forEach { (key, value) ->
                layer.putIfAbsent(key, value)
            }
        }
    }

    override fun toString(): String {
        return this.asSequence().joinToString(prefix = "[", postfix = "]") { element ->
            if (element is ByteArray) "${BigInteger(element)}" else "$element"
        }
    }

    override fun equals(other: Any?): Boolean {
        if (other is LayeredStack<*>) {
            if (other.size == this.size) {
                val ownIterator = asSequence().iterator()
                val otherIterator = other.asSequence().iterator()
                while (ownIterator.hasNext()) {
                    if (ownIterator.next() != otherIterator.next()) {
                        return false
                    }
                }
            }
        }
        return true
    }

    override fun hashCode(): Int {
        return size.hashCode()
    }

    fun asSequence() = sequence {
        for (i in 0 until this@LayeredStack.size) {
            yield(this@LayeredStack[i])
        }
    }

    companion object {
        var MINIMUM_LAYER_SIZE: Int = 6
    }
}
